crates/sui-core/src/state_accumulator.rs

```diff
@@ -0,0 +1,588 @@
+    highest_scheduled_seq_num: Option<CheckpointSequenceNumber>,
+}
+
+impl StateAccumulator {
+    pub fn new(
+        authority_state: Arc<AuthorityState>,
+        checkpoint_store: Arc<CheckpointStore>,
+        queue_size: u64,
+        store_path: &Path,
+    ) -> (Self, EnqueueHandle) {
+        let store = StateAccumulatorStore::new(store_path);
+        let (sender, receiver) = mpsc::channel(queue_size as usize);
+        let handle = EnqueueHandle {
+            sender,
+            store: store.clone(),
+        };
+
+        (
+            Self {
+                store,
+                authority_state,
+                checkpoint_store,
+                queue: receiver,
+                // Only saved for self enqueueing during crash recovery
+                sender_handle: handle.clone(),
+                latest_enqueued_state: None,
+                highest_scheduled_seq_num: None,
+            },
+            handle,
+        )
+    }
+
+    pub async fn run(&mut self, epoch: EpochId) {
+        self.handle_crash_recovery()
+            .await
+            .expect("StateAccumulator crash recovery cannot fail");
+
+        self.gen_checkpoint_state_hashes().await;
+        self.accumulate(epoch).await.expect("Accumulation failed");
+    }
+
+    /// To ensure consistency in the case of a crash after a checkpoint is enqueued
+    /// for hashing but before the checkpoint hash is committed, this function re-enqueues
+    /// all checkpoints between StateAccumulator::HighestHashed and
+    /// CheckpointExecutor::HighestExecuted. By doing this, we provide an eventual
+    /// consistency guarantee to components after the call to enqueue. Note that this
+    /// also works for checkpoint creation without any extra work, as checkpoints created
+    /// locally will still be considered by CheckpointExecutor. Lastly, for the special
+    /// case of creation of last checkpoint of the epoch, this will not be needed, as checkpoint
+    /// creation will block on enqueue, and then block waiting for the root state hash, before
+    /// moving on with consensus. If we fail during wait, CheckpointBuilder will try again.
+    ///
+    /// IMPORTANT: must be called before any external tasks attempt to enqueue, otherwise
+    /// may result in errors from enqueueing checkpoints out of order
+    async fn handle_crash_recovery(&self) -> SuiResult {
+        let next_to_hash = self
+            .store
+            .get_highest_hashed_checkpoint_seq_number()?
+            .map(|num| num.saturating_add(1))
+            .unwrap_or(0);
+        let last_to_hash = self
+            .checkpoint_store
+            .get_highest_executed_checkpoint_seq_number()?
+            .map(|num| num.saturating_add(1))
+            .unwrap_or(0);
+
+        for seq_num in next_to_hash..last_to_hash {
+            self.sender_handle
+                .enqueue(self.get_state_for_checkpoint_seq_num(seq_num))
```

Comment:
i wonder if this could just leverage the existing watermarks and crash recovery in CheckpointExecutor? It already has to ensure that every checkpoint is fully executed, so additionally it could ensure that every checkpoint has an accumulator stored (by calling into this module).

The exception would be final checkpoint creation (which is not driven by CheckpointExecutor), but that case already has its own crash recovery guarantees.

Reply:
We are using the same watermark as executor here. Both are stored in the checkpoint store. As far as the logic being moved to executor, that feels like it would break separation of concerns for executor to be re-enqueueing checkpoints that it already executed and moved on from during its startup. Moreover, this assumes that it's impossible for one to crash independent of the other. True today, but it would be nice to not have to assume that.

Lmk if I'm not correctly following your suggestion